### PR TITLE
Allow parsing of buffers

### LINKED ISSUE #10: Feature Request: fromBuffer
It's currently possible to convert to a buffer but not back again.

The simplest fromBuffer would ensure the length is either 16 or 4 then map those back to fields.

It would still have some properties, if was binary or string originally, v4, v6 or 4in6, etc.

toBuffer may want to reciprocate the same way toString does and return 4 or 16 but taking a format specified otherwise.

### PR DESCRIPTION
**Pull request by davewelsh**
Resolves #10

### CODE REVIEW

File: ip6addr.js [hunk 1]
```diff
@@ -683,6 +685,26 @@ function parseLong(input) {
   return out;
 }
 
+function parseBuffer(input) {
```
Comment (danmcd) [inline ip6addr.js:688]: assert.Buffer(input);  please?

File: test/parse.test.js [hunk 2]
```diff
@@ -105,3 +105,43 @@ test('parse long', function (t) {
   t.throws(pb(1.5), null, 'float throws');
   t.end();
 });
+
+test('parse buffer', function (t) {
+  t.ok(parse(0));
+  t.equal(parse(Buffer.from([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1 ])).toString(), '127.0.0.1');
```
Comment (danmcd) [inline test/parse.test.js:111]: No... this is wrong.  ::127.0.0.1 is NOT an IPv4 mapped address, it's literally ::127.0.0.1 or ::7f00:0.  You have other mistakes of this form.

File: test/parse.test.js [hunk 3]
```diff
@@ -105,3 +105,43 @@ test('parse long', function (t) {
...
+  t.equal(parse(Buffer.from([ 255, 128, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])).toString(), 'ff80:4000::');
+  t.equal(parse(Buffer.from([ 255, 128, 64, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])).toString(), 'ff80:4020::');
+
+  var strAddr = parse('1.2.3.4');
+  var bufferAddr = parse(Buffer.from([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4 ]));
+  t.equal(0, strAddr.compare(bufferAddr), 'parsing as buffer or long should be equivalent');
+
+  var ipv41 = parse(Buffer.from([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 192, 168, 0, 2 ]));
+  var ipv42 = parse(Buffer.from([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 192, 168, 0, 2 ]));
+  t.equal(0, ipv41.compare(ipv42), 'parsing a ipv4 representation of a buffer with bytes 11 and 12 equally 0x00 or 0xff should be equivalent');
```
Comment (danmcd) [inline test/parse.test.js:135]: This is completely wrong.  I can't review this further until you properly understand IPv6 addressing differences between ::<IPv4> and ::ffff:<IPv4>

File: ip6addr.js [hunk 4]
```diff
@@ -683,6 +685,26 @@ function parseLong(input) {
...
+  if (input.length !== 16) {
+    throw new Error('Buffer must have length of 16');
+  }
+  var out = new Addr();
+  for (i = 0; i < 8; i++) {
+    out._fields[i] = (input[i * 2] << 8) + input[(i * 2) + 1];  
+  }
+  
+  if (input.filter((_, i) => i < 10).every(v => v === 0)) { // the first 10 bytes are zero
+    if ((input[10] === 0 && input[11] === 0) || (input[10] === 255 && input[11] === 255)) {
```
Comment (danmcd) [inline ip6addr.js:698]: This is fundamentally wrong.  You only do this for the ::ffff:<IPv4> case.

**Review by danmcd — CHANGES_REQUESTED** [review]
Please revisit your understanding of how IPv6 addresses work, and then update your code and tests accordingly.

**Comment by danmcd** [comment]
Two more big-picture things:

1.) This change will need a version bump ("0.2.4" I'm thinking).

2.) Make sure "gmake check" succeeds.

File: test/parse.test.js (hunk 2 above)
Comment (davewelsh) [reply test/parse.test.js:111]: My intention was to try to keep the same kind of functionality the existing code has.

E.g. `ip6addr.parse('127.0.0.1').toBuffer()` returns `<Buffer 00 00 00 00 00 00 00 00 00 00 ff ff c0 a8 00 01>`. So `ip6addr.parse(Buffer.from(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 0, 0, 1)).toString()` should return the orginal `'127.0.0.1'`.

It also seems that `ip6addr.parse(Buffer.from(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1)).toString()` should return `'127.0.0.1'`.

Perhaps if the original `ip6addr.parse('127.0.0.1').toBuffer()` returned a 4-byte buffer `<Buffer c0 a8 00 01>`, then the buffer-parsing function could be modified to distinguish between true ipv4 and mapped ipv4 addresses, but I didn't want to modify any of the existing functionality.

I'm not sure there's a path forward for this PR, but thanks for your work in considering it!

Comment (davewelsh) [reply test/parse.test.js:111]: I reexamined the internal data structure for `parse('127.0.0.1')`, `parse('::ffff:127.0.0.1') and `parse('::ffff:7f00:01')`.

They are like this:

**ip.parse('127.0.0.1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Bare: true, ipv4Mapped: true } }
```

**ip.parse('::ffff:127.0.0.1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Mapped: true } }
```

**ip.parse('::ffff:7f00:1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: {} }
```

A buffer-parsing function could avoid setting any `_attrs`. I.e., there might not be a good reason to treat `Buffer<00 00 00 00 00 00 00 00 00 00 ff ff, 7f 00 00 01>` as a mapped ipv4 address as opposed to a plain ipv6 address.

Would that be preferred?

Then `parse(parse('::ffff:7f00:1').toBuffer()).toString()` would equal `'::ffff:7f00:1'`, but `parse(parse('127.0.0.1').toBuffer()).toString()` would also equal `'::ffff:7f00:1'`.

Comment (danmcd) [reply test/parse.test.js:111]: There is a difference between '127.0.0.1' and '::127.0.0.1'.  The first is IPv4, which gets converted into '::ffff:127.0.0.1'.  The second is native IPv6 (albeit not used) and should not be converted to any sort of IPv4, whether mapped ('::ffff:<v4>') or native IPv4.

File: test/parse.test.js (hunk 3 above)
Comment (danmcd) [reply test/parse.test.js:135]: To elaborate:  ipv41's example should convert to '::192.168.0.2' not '192.168.0.2'.

File: test/parse.test.js (hunk 2 above)
Comment (danmcd) [reply test/parse.test.js:111]: To elaborate, the buffer data in line 111 should result in '::127.0.0.1' not '127.0.0.1', aka. '::ffff:127.0.0.1'.

**Review by danmcd — CHANGES_REQUESTED** [review]
Remember that '::IPv4' is not the same as 'IPv4'.  '::ffff:IPv4' is the same as 'IPv4'.

File: test/parse.test.js (hunk 3 above)
Comment (davewelsh) [reply test/parse.test.js:135]: >To elaborate: ipv41's example should convert to '::192.168.0.2' not '192.168.0.2'.

Would you prefer it to return '::c0a8:2' instead? `parse('::192.168.0.2')` currently throws an error.

Comment (danmcd) [reply test/parse.test.js:135]: Really?!? No wonder it wasn't properly tested. For now, let's keep things least-surprise and indeed have it print '::c0a8:2'.

File: test/parse.test.js (hunk 2 above)
Comment (davewelsh) [reply test/parse.test.js:111]: I agree. `Buffer<00 00 00 00 00 00 00 00 00 00 00 00 7f 00 00 01>` should not turn into '127.0.0.1' and should stay '::127.0.01' (or rather '::f700:1', based the other discussion). The way forward seems to be to not modify _field bytes 10 and 11 and to not set any _attr keys.

---

For Buffers like `Buffer<00 00 00 00 00 00 00 00 00 00 ff ff 7f 00 00 01>`, we seem to have a few options though:

The `Addr` created by `parse('127.0.0.1')`, `parse('::ffff:127.0.0.1')`, and `parse('::ffff:f700:1')` will all give this buffer when `toBuffer` is called. So which type of `Addr` should be created when parsing such a buffer?

1. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ip4Bare: true, ipv4Mapped: true } }` (what is created by `parse('127.0.0.1')`)
2. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Mapped: true } }` (what is created by `parse('::ffff:127.0.0.1')`)
3. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { } }` (what is created by `parse('::ffff:7f00:1')`)

My inclination is 3, i.e., not setting any keys of _attrs, but I think any of these options is reasonable.

Comment (danmcd) [reply test/parse.test.js:111]: Your question is an interesting one, and I need to know "Where are you getting these Buffers from"? In an IPv6 packet you'll never see v4mapped, but some protocols may use IPv6 address space to express both.  Looking at what you feed via parse(), I'd go with option 2.

My concern this whole time has been you're reading '00 00 00 00 00 00 00 00 00 00 00 00 7f 00 00 01' incorrectly. '00 00 00 00 00 00 00 00 00 00 ff ff aaa bbb ccc ddd' should always become an IPv4 mapped address.

**Comment by davewelsh** [comment]
> Two more big-picture things:
> 
> 1.) This change will need a version bump ("0.2.4" I'm thinking).
> 
> 2.) Make sure "gmake check" succeeds.

Should I take it upon myself to change the version number to 0.2.4 in this PR?

I've installed gnu make, but it seems to have an issue with the version of Python I have installed (3.9.5):

```
$ make check
cd deps/javascriptlint && make install
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'sys' has no attribute 'maxint'
make[1]: Entering directory `C:/Users/DaveWelsh/node-ip6addr/deps/javascriptlint'
(cd spidermonkey/src && CC="cc" CFLAGS="-fno-strict-aliasing -O -fPIC -m" C:/make/bin/make)
make[2]: Entering directory `C:/Users/DaveWelsh/node-ip6addr/deps/javascriptlint/spidermonkey/src'
cc -fno-strict-aliasing -O -fPIC -m -Wall -Wno-format -O -DXP_UNIX -O -DXP_UNIX -DSVR4 -DSYSV -DSOLARIS -DHAVE_LOCALTIME_R -UDEBUG -DNDEBUG -UDEBUG_dap -DEDITLINE -Ibuild -o build/jscpucfg.o -c jscpucfg.c
process_begin: CreateProcess(NULL, cc -fno-strict-aliasing -O -fPIC -m -Wall -Wno-format -O -DXP_UNIX -O -DXP_UNIX -DSVR4 -DSYSV -DSOLARIS -DHAVE_LOCALTIME_R -UDEBUG -DNDEBUG -UDEBUG_dap -DEDITLINE -Ibuild -o build/jscpucfg.o -c jscpucfg.c, ...) failed.
make (e=2): The system cannot find the file specified.
make[2]: *** [build/jscpucfg.o] Error 2
make[2]: Leaving directory `C:/Users/DaveWelsh/node-ip6addr/deps/javascriptlint/spidermonkey/src'
make[1]: *** [spidermonkey/src/build/libjs.a] Error 2
make[1]: Leaving directory `C:/Users/DaveWelsh/node-ip6addr/deps/javascriptlint'
make: *** [deps/javascriptlint/build/install/jsl] Error 2
```

**Pull request closed without merging**